### ISSUE TITLE
How to get around 69 exact_match performance?

### ISSUE THREAD

**Issue by duyvuleo** [ElementAI/duorat#3]
Hi authors,

Thanks for the excellent work. I really struggled to use RAT-SQL code (from Microsoft) to get >60 exact_match performance (its training was really unstable). 

Switching to DuoRAT saves me a lot of time. 

For the initial results, I trained DuoRAT with two configurations (duorat-finetune-bert-large & duorat-new-db-content) with only slight modification on batch_size=4 (instead of 9) due to the OOM issue on my GPU (16GB V100). 

For duorat-finetune-bert-large, after 55000 steps, I got:

```
[2020-10-28T08:01:05] Step 5000 stats, val: loss = 0.12590350662727393, easy_exact = 0.7620967741935484, medium_exact = 0.49327354260089684, hard_exact = 0.40804597701149425, extra_exact = 0.2469879518072289, all_exact = 0.5038684719535783
[2020-10-28T09:22:16] Step 10000 stats, val: loss = 0.14725026781031253, easy_exact = 0.7782258064516129, medium_exact = 0.594170403587444, hard_exact = 0.5172413793103449, extra_exact = 0.3614457831325301, all_exact = 0.5880077369439072
[2020-10-28T10:44:27] Step 15000 stats, val: loss = 0.14249609349245057, easy_exact = 0.842741935483871, medium_exact = 0.6547085201793722, hard_exact = 0.5, extra_exact = 0.3253012048192771, all_exact = 0.620889748549323
[2020-10-28T12:04:45] Step 20000 stats, val: loss = 0.15154095876752985, easy_exact = 0.8306451612903226, medium_exact = 0.6636771300448431, hard_exact = 0.5, extra_exact = 0.3795180722891566, all_exact = 0.6305609284332688
[2020-10-28T13:26:16] Step 25000 stats, val: loss = 0.15309250173276034, easy_exact = 0.8790322580645161, medium_exact = 0.6838565022421524, hard_exact = 0.5574712643678161, extra_exact = 0.3674698795180723, all_exact = 0.6586073500967118
[2020-10-28T14:46:39] Step 30000 stats, val: loss = 0.15393706009051325, easy_exact = 0.8548387096774194, medium_exact = 0.6053811659192825, hard_exact = 0.47126436781609193, extra_exact = 0.3433734939759036, all_exact = 0.6005802707930368
[2020-10-28T16:06:07] Step 35000 stats, val: loss = 0.16426039838576167, easy_exact = 0.8669354838709677, medium_exact = 0.6771300448430493, hard_exact = 0.4942528735632184, extra_exact = 0.3433734939759036, all_exact = 0.6382978723404256
[2020-10-28T17:26:28] Step 40000 stats, val: loss = 0.1535649892186829, easy_exact = 0.8467741935483871, medium_exact = 0.647982062780269, hard_exact = 0.5229885057471264, extra_exact = 0.3373493975903614, all_exact = 0.6247582205029013
[2020-10-28T18:46:05] Step 45000 stats, val: loss = 0.16708454437467415, easy_exact = 0.8629032258064516, medium_exact = 0.6636771300448431, hard_exact = 0.5057471264367817, extra_exact = 0.3373493975903614, all_exact = 0.632495164410058
[2020-10-28T20:05:48] Step 50000 stats, val: loss = 0.1519994481004652, easy_exact = 0.8629032258064516, medium_exact = 0.625560538116592, hard_exact = 0.5229885057471264, extra_exact = 0.3433734939759036, all_exact = 0.6199226305609284
[2020-10-28T21:25:47] Step 55000 stats, val: loss = 0.17041306306087417, easy_exact = 0.8306451612903226, medium_exact = 0.6614349775784754, hard_exact = 0.5229885057471264, extra_exact = 0.35542168674698793, all_exact = 0.6295938104448743
```

For duorat-new-db-content, I got:

```
[2020-10-28T10:05:47] Step 5000 stats, val: loss = 0.12354135784776286, easy_exact = 0.7983870967741935, medium_exact = 0.5650224215246636, hard_exact = 0.42528735632183906, extra_exact = 0.25301204819277107, all_exact = 0.5473887814313346
[2020-10-28T11:33:05] Step 10000 stats, val: loss = 0.14531567805974493, easy_exact = 0.8306451612903226, medium_exact = 0.5986547085201793, hard_exact = 0.5172413793103449, extra_exact = 0.22289156626506024, all_exact = 0.5802707930367504
[2020-10-28T12:53:33] Step 15000 stats, val: loss = 0.15475997000768765, easy_exact = 0.8387096774193549, medium_exact = 0.6681614349775785, hard_exact = 0.4827586206896552, extra_exact = 0.3674698795180723, all_exact = 0.6295938104448743
[2020-10-28T14:18:32] Step 20000 stats, val: loss = 0.16852696785951896, easy_exact = 0.8064516129032258, medium_exact = 0.6412556053811659, hard_exact = 0.4942528735632184, extra_exact = 0.3674698795180723, all_exact = 0.6121856866537717
[2020-10-28T15:39:26] Step 25000 stats, val: loss = 0.1531622735157472, easy_exact = 0.7741935483870968, medium_exact = 0.6143497757847534, hard_exact = 0.5229885057471264, extra_exact = 0.39759036144578314, all_exact = 0.6025145067698259
[2020-10-28T16:59:41] Step 30000 stats, val: loss = 0.14458048883357177, easy_exact = 0.8467741935483871, medium_exact = 0.647982062780269, hard_exact = 0.5517241379310345, extra_exact = 0.37349397590361444, all_exact = 0.6353965183752418
[2020-10-28T18:20:00] Step 35000 stats, val: loss = 0.14305388022302443, easy_exact = 0.8669354838709677, medium_exact = 0.7174887892376681, hard_exact = 0.4885057471264368, extra_exact = 0.41566265060240964, all_exact = 0.6663442940038685
[2020-10-28T19:39:42] Step 40000 stats, val: loss = 0.1363988071367076, easy_exact = 0.8911290322580645, medium_exact = 0.7040358744394619, hard_exact = 0.5402298850574713, extra_exact = 0.3493975903614458, all_exact = 0.6644100580270793
[2020-10-28T20:59:41] Step 45000 stats, val: loss = 0.1358668204875234, easy_exact = 0.8709677419354839, medium_exact = 0.6345291479820628, hard_exact = 0.4827586206896552, extra_exact = 0.3855421686746988, all_exact = 0.625725338491296
[2020-10-28T22:19:24] Step 50000 stats, val: loss = 0.14011963381750844, easy_exact = 0.8790322580645161, medium_exact = 0.679372197309417, hard_exact = 0.5, extra_exact = 0.3433734939759036, all_exact = 0.6431334622823984
```

Do they look correct? Do I need to wait for it to be trained for a bit longer to get around 69 performance? 

Thanks!

**Comment by duyvuleo** [comment]
Also, during training, I observed that:
```
[2020-10-29T00:51:05] Step 50 stats, train: step = 50
[2020-10-29T00:51:05] Step 50: loss=1.4494
[2020-10-29T00:51:05] Step 50: lr[no-bert]=0.0000
[2020-10-29T00:51:05] Step 50: lr[bert]=0.0000
```

I was wondering why learning rates are all zeros. Can everyone explain it a bit? Thanks!

**Comment by tscholak** [comment]
Hi @duyvuleo, I'm glad you find our code useful!
Please find attached a plot of exact match % on val against training steps / 5000 for a job that achieved about 70% in the end:
<img width="1077" alt="Screen Shot 2020-10-28 at 9 16 23 PM" src="https://user-images.githubusercontent.com/1568873/97513543-eedecb80-1962-11eb-8bf3-8904a15ab198.png">
Your numbers are not unusual, you should try running your jobs for the full 100,000 steps.
However, I suspect that the reduction to `batch_size=4` will make reaching peak performance difficult. We found that `batch_size=9` is really the lowest value needed for getting good results. If you decrease the batch size, you need to increase the number of gradient accumulation steps to compensate. In the default config we use 3 accumulation steps. For your setting, you may need 7 to keep the effective batch size somewhat constant.
With regards to the learning rate, this is normal. Only four decimal places are printed for some reason. Here's a plot of the learning rate as a function of training steps / 10:
<img width="1141" alt="Screen Shot 2020-10-28 at 9 25 50 PM" src="https://user-images.githubusercontent.com/1568873/97514036-2ac66080-1964-11eb-91be-088199f5d9f4.png">
I hope this helps, let me know if you have further questions.

**Comment by duyvuleo** [comment]
Thanks for your prompt response. 

I re-run some configurations with your suggestions regarding batch size and gradient accumulation steps. Hope I can replicate ~70 exact_match performance.

**Comment by tscholak** [comment]
Good luck, it should work 😀
Btw, I saw that you maintain a fork of duorat where you have made some changes to support non-docker operation. Feel free to clean this up a bit and open a PR to upstream. We accept contributions 😊

**Comment by duyvuleo** [comment]
Hi @tscholak 
I let the previous run to continue until 100K steps and managed to get 68.6 performance. I think it should be good.
I also re-run with batch_size=4 (gradient accumulation steps=7) and after 20K steps I got ~68.1. Hopefully, it can reach 69-70 at 100K steps.    

Sure.  Thanks! I am very interested in and would love to contribute to this repo. I will be submitting some PRs shortly.

**Comment by tscholak** [comment]
I suppose you were successful? Nice 👍

**Comment by duyvuleo** [comment]
Yeah. I got ~72 exact_match performance which looks nice.

**Comment by JoaoLages** [comment]
can someone share a pretrained model please? :)

**Comment by awasthiabhijeet** [comment]
> Good luck, it should work 
> Btw, I saw that you maintain a fork of duorat where you have made some changes to support non-docker operation. Feel free to clean this up a bit and open a PR to upstream. We accept contributions

Hi @duyvuleo ,
I am trying to set-up this project on a shared server.
Using docker in a root-less mode seems a bit of a stretch.
Could you please point to the changes which you made to support non-docker operation?

Thanks,
